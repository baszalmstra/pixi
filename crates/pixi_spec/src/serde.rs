use rattler_conda_types::{
    BuildNumberSpec, NamedChannelOrUrl, ParseStrictness::Strict, StringMatcher, VersionSpec,
};
use rattler_digest::{Md5Hash, Sha256Hash};
use serde::{de::Error, Deserialize, Deserializer, Serialize, Serializer};
use serde_with::serde_as;
use url::Url;

use crate::{DetailedVersionSpec, GitRev, GitSpec, PathSpec, Spec, UrlSpec};

#[serde_as]
#[derive(serde::Deserialize)]
#[serde(deny_unknown_fields)]
struct RawSpec {
    /// The version spec of the package (e.g. `1.2.3`, `>=1.2.3`, `1.2.*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub version: Option<VersionSpec>,

    /// The URL of the package
    pub url: Option<Url>,

    /// The git url of the package
    pub git: Option<Url>,

    /// The path to the package
    pub path: Option<String>,

    /// The git revision of the package
    pub branch: Option<String>,

    /// The git revision of the package
    pub rev: Option<String>,

    /// The git revision of the package
    pub tag: Option<String>,

    /// The build string of the package (e.g. `py37_0`, `py37h6de7cb9_0`, `py*`)
    #[serde_as(as = "Option<serde_with::DisplayFromStr>")]
    pub build: Option<StringMatcher>,

    /// The build number of the package
    pub build_number: Option<BuildNumberSpec>,

    /// Match the specific filename of the package
    pub file_name: Option<String>,

    /// The channel of the package
    pub channel: Option<NamedChannelOrUrl>,

    /// The subdir of the channel
    pub subdir: Option<String>,

    /// The md5 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Md5>>")]
    pub md5: Option<Md5Hash>,

    /// The sha256 hash of the package
    #[serde_as(as = "Option<rattler_digest::serde::SerializableHash::<rattler_digest::Sha256>>")]
    pub sha256: Option<Sha256Hash>,
}

impl<'de> Deserialize<'de> for Spec {
    fn deserialize<D>(deserializer: D) -> Result<Spec, D::Error>
    where
        D: Deserializer<'de>,
    {
        serde_untagged::UntaggedEnumVisitor::new()
            .expecting(
                "a version string like \">=0.9.8\" or a detailed dependency like { version = \">=0.9.8\" }",
            )
            .string(|str| {
                VersionSpec::from_str(str, Strict)
                    .map_err(serde_untagged::de::Error::custom)
                    .map(Spec::Version)
            })
            .map(|map| {
                let raw_spec: RawSpec = map.deserialize()?;

                if raw_spec.git.is_none()
                    && (raw_spec.branch.is_some()
                        || raw_spec.rev.is_some()
                        || raw_spec.tag.is_some())
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`branch`, `rev`, and `tag` are only valid when `git` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.build.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`build` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.build_number.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`build-number` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.file_name.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`file-name` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.channel.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`channel` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.subdir.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`subdir` is only valid when `version` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.url.is_none() && raw_spec.sha256.is_some()
                {
                    return Err(serde_untagged::de::Error::custom(
                        "`sha256` is only valid when `version` or `url` is specified",
                    ));
                }

                if raw_spec.version.is_none() && raw_spec.url.is_none() && raw_spec.md5.is_some() {
                    return Err(serde_untagged::de::Error::custom(
                        "`md5` is only valid when `version` or `url` is specified",
                    ));
                }

                let spec = match (raw_spec.version, raw_spec.url, raw_spec.path, raw_spec.git) {
                    (Some(version), None, None, None) => Spec::DetailedVersion(DetailedVersionSpec {
                        version,
                        build: raw_spec.build,
                        build_number: raw_spec.build_number,
                        file_name: raw_spec.file_name,
                        channel: raw_spec.channel,
                        subdir: raw_spec.subdir,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, Some(url), None, None) => Spec::Url(UrlSpec {
                        url,
                        md5: raw_spec.md5,
                        sha256: raw_spec.sha256,
                    }),
                    (None, None, Some(path), None) => Spec::Path(PathSpec { path: path.into() }),
                    (None, None, None, Some(git)) => {
                        let rev = match (raw_spec.branch, raw_spec.rev, raw_spec.tag) {
                            (Some(branch), None, None) => Some(GitRev::Branch(branch)),
                            (None, Some(rev), None) => Some(GitRev::Commit(rev)),
                            (None, None, Some(tag)) => Some(GitRev::Tag(tag)),
                            (None, None, None) => None,
                            _ => {
                                return Err(serde_untagged::de::Error::custom(
                                    "only one of `branch`, `rev`, or `tag` can be specified",
                                ));
                            }
                        };
                        Spec::Git(GitSpec { git, rev })
                    }
                    (None, None, None, None) => {
                        return Err(serde_untagged::de::Error::custom(
                            "one of `version`, `url`, or `path` must be specified",
                        ))
                    }
                    (_, _, _, _) => {
                        return Err(serde_untagged::de::Error::custom(
                            "only one of `version`, `url`, `path`, or `git` can be specified",
                        ))
                    }
                };

                Ok(spec)
            })
            .deserialize(deserializer)
    }
}

impl<'de> Deserialize<'de> for PathSpec {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        struct Raw {
            path: String,
        }

        Raw::deserialize(deserializer).map(|raw| PathSpec {
            path: raw.path.into(),
        })
    }
}

impl Serialize for PathSpec {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        #[derive(Serialize)]
        struct Raw {
            path: String,
        }

        Raw {
            path: self.path.to_string(),
        }
        .serialize(serializer)
    }
}

#[cfg(test)]
mod test {
    use serde::Serialize;
    use serde_json::{json, Value};

    use super::*;

    #[test]
    fn test_round_trip() {
        let examples = [
            json! { "1.2.3" },
            json!({ "version": "1.2.3" }),
            json!({ "ver": "1.2.3" }),
            json! { "*" },
            json!({ "path": "foobar" }),
            json!({ "path": "foobar", "version": "1.2.3" }),
            json!({ "version": "//" }),
            json!({ "path": "foobar", "version": "//" }),
            json!({ "path": "foobar", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "version": "1.2.3", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda" }),
            json!({ "url": "https://conda.anaconda.org/conda-forge/linux-64/21cmfast-3.3.1-py38h0db86a8_1.conda", "sha256": "315f5bdb76d078c43b8ac0064e4a0164612b1fce77c869345bfc94c75894edd3" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main" }),
            json!({ "git": "https://github.com/conda-forge/21cmfast-feedstock", "branch": "main", "tag": "v1" }),
        ];

        #[derive(Serialize)]
        struct Snapshot {
            input: Value,
            result: Value,
        }

        let mut snapshot = Vec::new();
        for input in examples {
            let spec: Result<Spec, _> = serde_json::from_value(input.clone());
            let result = match spec {
                Ok(spec) => {
                    let spec = Spec::from(spec);
                    serde_json::to_value(&spec).unwrap()
                }
                Err(e) => {
                    json!({
                        "error": format!("ERROR: {e}")
                    })
                }
            };

            snapshot.push(Snapshot { input, result });
        }

        insta::assert_yaml_snapshot!(snapshot);
    }
}
