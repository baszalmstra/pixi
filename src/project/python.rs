use crate::utils::spanned::PixiSpanned;
use indexmap::IndexMap;
use serde::de::{DeserializeSeed, Error, MapAccess, Visitor};
use serde::{Deserialize, Deserializer};
use std::collections::HashMap;
use std::fmt::Formatter;

/// Represents a set of python dependencies on which a project can depend. The dependencies are
/// formatted using a modern version specifier. See [`pep508_rs::modern::RequirementModern`].
#[derive(Default, Debug, Clone)]
pub struct PypiDependencies {
    requirements: IndexMap<
        PixiSpanned<rip::PackageName>,
        PixiSpanned<(pep508_rs::modern::RequirementModern, pep508_rs::Requirement)>,
    >,
}

impl PypiDependencies {
    /// Returns `true` if no requirements have been specified
    pub fn is_empty(&self) -> bool {
        self.requirements.is_empty()
    }

    /// Returns the requirements as [`pep508_rs::Requirement`]s.
    pub fn as_pep508(&self) -> Vec<pep508_rs::Requirement> {
        self.requirements
            .values()
            .map(|s| s.value.1.clone())
            .collect()
    }
}

impl<'de> Deserialize<'de> for PypiDependencies {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        struct RequirementsVisitor;
        struct RequirementVisitor<'i> {
            name: &'i str,
        }

        impl<'de, 'i> DeserializeSeed<'de> for RequirementVisitor<'i> {
            type Value =
                PixiSpanned<(pep508_rs::modern::RequirementModern, pep508_rs::Requirement)>;

            fn deserialize<D>(self, deserializer: D) -> Result<Self::Value, D::Error>
            where
                D: Deserializer<'de>,
            {
                let PixiSpanned { value, span } =
                    PixiSpanned::<pep508_rs::modern::RequirementModern>::deserialize(deserializer)?;
                let req = value
                    .to_pep508(self.name, &HashMap::default())
                    .map_err(D::Error::custom)?;
                Ok(PixiSpanned {
                    value: (value, req),
                    span,
                })
            }
        }

        impl<'de> Visitor<'de> for RequirementsVisitor {
            type Value = IndexMap<
                PixiSpanned<rip::PackageName>,
                PixiSpanned<(pep508_rs::modern::RequirementModern, pep508_rs::Requirement)>,
            >;

            fn expecting(&self, formatter: &mut Formatter) -> std::fmt::Result {
                formatter.write_str("a mapping from package names to requirements")
            }

            fn visit_map<A>(self, mut map: A) -> Result<Self::Value, A::Error>
            where
                A: MapAccess<'de>,
            {
                let mut result = IndexMap::new();
                while let Some(key) = map.next_key::<PixiSpanned<rip::PackageName>>()? {
                    let value = map.next_value_seed(RequirementVisitor {
                        name: key.value.as_str(),
                    })?;
                    result.insert(key, value);
                }
                Ok(result)
            }
        }

        deserializer
            .deserialize_map(RequirementsVisitor {})
            .map(|requirements| PypiDependencies { requirements })
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use insta::assert_debug_snapshot;

    #[test]
    fn test_only_version() {
        let requirement: PypiDependencies = toml_edit::de::from_str("foo = \">=3.12\"").unwrap();
        assert_debug_snapshot!(requirement);
    }

    #[test]
    fn test_extended() {
        let requirement: PypiDependencies =
            toml::de::from_str("foo = { version=\">=3.12\", extras = [\"bar\"] }").unwrap();
        assert_debug_snapshot!(requirement);
    }

    #[test]
    fn test_invalid_git_dependency_error() {
        let requirement = toml::de::from_str::<PypiDependencies>(
            "foo = { git=\"https://github.com/foo/bar\", branch = \"main\", rev=\"deadbeef\" }",
        );
        assert_debug_snapshot!(requirement.unwrap_err());
    }

    // TODO: This crashes currently but should instead result in an error. This should be fixed in pep508_rs.
    //#[test]
    // fn test_invalid_extra() {
    //     let requirement = toml::de::from_str::<PythonDependencies>(
    //         r#"foo = { version="3.2.1", extras=["b$r"] }"#,
    //     );
    //     assert_debug_snapshot!(requirement.unwrap_err());
    // }

    // TODO: pep508_rs does not handle ^ operator yet.
}
