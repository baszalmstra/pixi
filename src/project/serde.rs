use rattler_conda_types::{Channel, ChannelConfig};
use serde::de::Error;
use serde::{Deserialize, Deserializer};
use std::borrow::Cow;

pub struct ChannelStr;

impl<'de> serde_with::DeserializeAs<'de, Channel> for ChannelStr {
    fn deserialize_as<D>(deserializer: D) -> Result<Channel, D::Error>
    where
        D: Deserializer<'de>,
    {
        let channel_str = Cow::<str>::deserialize(deserializer)?;
        let channel_config = ChannelConfig::default();
        Channel::from_str(channel_str, &channel_config).map_err(D::Error::custom)
    }
}
