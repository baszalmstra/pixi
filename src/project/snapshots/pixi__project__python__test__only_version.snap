---
source: src/project/python.rs
assertion_line: 102
expression: requirement
---
PypiDependencies {
    requirements: {
        PixiSpanned {
            span: Some(
                0..3,
            ),
            value: PackageName {
                source: "foo",
                normalized: "foo",
            },
        }: PixiSpanned {
            span: Some(
                6..14,
            ),
            value: (
                Dependency(
                    VersionSpecifier(
                        VersionSpecifiers(
                            [
                                VersionSpecifier {
                                    operator: GreaterThanEqual,
                                    version: Version {
                                        epoch: 0,
                                        release: [
                                            3,
                                            12,
                                        ],
                                        pre: None,
                                        post: None,
                                        dev: None,
                                        local: None,
                                    },
                                },
                            ],
                        ),
                    ),
                ),
                Requirement {
                    name: "foo",
                    extras: None,
                    version_or_url: Some(
                        VersionSpecifier(
                            VersionSpecifiers(
                                [
                                    VersionSpecifier {
                                        operator: GreaterThanEqual,
                                        version: Version {
                                            epoch: 0,
                                            release: [
                                                3,
                                                12,
                                            ],
                                            pre: None,
                                            post: None,
                                            dev: None,
                                            local: None,
                                        },
                                    },
                                ],
                            ),
                        ),
                    ),
                    marker: None,
                },
            ),
        },
    },
}
